src/semantics.js

```diff
@@ -0,0 +1,93 @@
+/* eslint-disable unicorn/no-null */
```

Comment:
NTH.

(I think I saw a message from @adamalton  to suggest this)
It'd be really nice to have a more declarative and generalised way to define sematics.